### ISSUE TITLE
subsequent runs of recipe in same buildout gets environment from previous run

### ISSUE THREAD

**Issue by lukenowak** [hexagonit/hexagonit.recipe.cmmi#1]
For example:

[buildout]
parts =
  compile1
  compile2

[compile1]
recipe = hexagonit.recipe.cmmi
environment =
  FLAG_WHICH_KILL = abcde
url = url1

[compile2]
recipe = hexagonit.recipe.cmmi
url = url2

While compile2 is run in same buildout run, environment variable FLAG_WHICH_KILL is accessible in os.environ['FLAG_WHICH_KILL']

Expected behaviour is that each part using h.r.cmmi would be run with separated environment.

**Comment by lukenowak** [comment]
This patch:

diff --git a/hexagonit/recipe/cmmi/**init**.py b/hexagonit/recipe/cmmi/**init**.py
index b46364d..783b1a2 100644
--- a/hexagonit/recipe/cmmi/**init**.py
+++ b/hexagonit/recipe/cmmi/**init**.py
@@ -91,9 +91,11 @@ class Recipe(object):
         patches = self.options.get('patches', '').split()

```
     if self.environ:
```
-            stored_envrion = {}
           for key in sorted(self.environ.keys()):
               log.info('[ENV] %s = %s', key, self.environ[key])
-            os.environ.update(self.environ)
-                stored_envrion[key] = os.environ.get(key)
- ```
             os.environ[key] = self.environ[key]
  
   # Download the source using hexagonit.recipe.download
   if self.options['url']:
  ```
  
  @@ -156,6 +158,12 @@ class Recipe(object):
                         'you can inspect what went wrong' % os.getcwd())
               raise
       finally:
-            if self.environ:
-                for key,value in stored_envrion.iteritems():
-                    if value is None:
-                        os.environ.pop(key, None)
-                    else:
- ```
                     os.environ[key] = value
       os.chdir(current_dir)
  
   if self.options['url']:
  ```

Fixes issue.

If such solution is approved I can play more with git and add test, please let me know

**Comment by hexagonit** [comment]
I agree with the problem description; the environment overrides should not leak between parts. The rendering of the diff above makes it hard to read, but perhaps simply making a copy of the original environment before modifying it (before = os.environ.copy()) and then reverting back to it (os.environ.clear(); os.environ.update(before)) in a finally block would do it.

**Comment by hexagonit** [comment]
Hi Luke

Could you check if 96b9fe5bd9836856c89d5ddf93e1b639a871f1d0 fixes the issue for you?

**Comment by lukenowak** [comment]
Yes, it fixes my issue. Thank you very much.

**Comment by fdiary** [comment]
I still got this problem with python-2.4 and I found that we should use 'Popen(cmd, shell=True, env=os.environ)' instead of 'os.system(cmd)' to avoid python-2.4 bug.

Please see the following snippet.

<pre>
$ cat 1.py
import os
b = os.environ.copy()
os.environ['XXX']='1'
os.system('env|grep XXX')
os.environ.clear()
os.environ.update(b)
os.system('env|grep XXX')
$ python2.4 1.py
XXX=1
XXX=1
$ python2.6 1.py
XXX=1
</pre>


It should be python-2.4's bug, but if we use 'Popen(cmd, shell=True, env=os.environ)' instead, we can avoid this bug, and anyway using Popen is preferable according to Python's documentation.

<pre>
$ cat 2.py
import os
from subprocess import Popen
b = os.environ.copy()
os.environ['XXX']='1'
proc = Popen('env|grep XXX', shell=True, env=os.environ)
os.waitpid(proc.pid, 0)
os.environ.clear()
os.environ.update(b)
proc = Popen('env|grep XXX', shell=True, env=os.environ)
os.waitpid(proc.pid, 0)
$ python2.4 2.py
XXX=1
$ python2.6 2.py
XXX=1
</pre>


So could you please apply the following patch?

<pre>
--- hexagonit/recipe/cmmi/__init__.py        2010-12-11 15:31:47.000000000 +0100
+++ hexagonit/recipe/cmmi/__init__.py        2010-12-11 19:19:19.000000000 +0100
@@ -4,6 +4,7 @@
 import logging
 import os
 import shutil
+from subprocess import Popen
 import zc.buildout
 
 class Recipe(object):
@@ -72,8 +73,10 @@
 
     def run(self, cmd):
         log = logging.getLogger(self.name)
-        if os.system(cmd):
-            log.error('Error executing command: %s' % cmd)
+        process = Popen(cmd, shell=True, env=os.environ)
+        status = os.waitpid(process.pid, 0)[1]
+        if status:
+            log.error('Error executing command: %s' % status)
             raise zc.buildout.UserError('System error')
 
     def is_build_dir(self):
</pre>


# BTW, if we use Popen with env argument, we no longer need to modify os.environ...

Thanks in advance,
Kazuhiko

**Comment by hexagonit** [comment]
Thanks for the report, I will review the patch and get back to you. Just for reference, is http://bugs.python.org/issue3227 the issue that you referred to?

cheers,
Kai

**Comment by fdiary** [comment]
Hello Kai,
Thanks for you reply.
Indeed, this issue seems to be the cause of my problem on Python-2.4.
Kazuhiko

**Comment by hexagonit** [comment]
I pushed a fix based on your patch in 8308537a78f9837413744164246644e1c17a1a2b. Could you please test if this fixes the issue for you.

I ended up using `subprocess.call()` since it handles the wait() logic automatically and also passing the a modified copy of `os.environ` in the call. However, this changes the environment handling for the hook scripts (they previously had access to the modified environment, but do not anymore) so I'm still contemplating whether this is acceptable.

cheers,
Kai

**Comment by fdiary** [comment]
Thanks for your commit. It works quite fine on my python-2.4 environment.
(And for me, the change of the environment handling for the hook scripts is acceptable.)

**Comment by hexagonit** [comment]
Thanks. I'll get some internal feedback here and push out a new version this week if everybody is satisfied with the change.

cheers,
Kai

**Comment by hexagonit** [comment]
Commit 97b6bc7f3f55da6f921656a95ed400d97def0f24 adds support for accessing the augmented environment in the hook scripts also.

If no bugs come up in the next few days I'll prepare a new release.

cheers,
Kai

**Comment by hexagonit** [comment]
I released version 1.5.0 today which contains the fixes.
- Kai
